llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp

```diff
@@ -55,15 +56,91 @@ void InterBlockScheduling::markEpilogueBlocks() {
   }
 }
 
+std::map<MachineBasicBlock *, LivePhysRegs> calculateLiveOuts(MachineFunction *MF) {
+  std::queue<MachineBasicBlock *> WorkList;
+  std::set<MachineBasicBlock *> WorkListSet;
+  std::map<MachineBasicBlock *, LivePhysRegs> LiveIns;
+  std::map<MachineBasicBlock *, LivePhysRegs> LiveOuts;
+  const TargetRegisterInfo *TRI = MF->getSubtarget().getRegisterInfo();
+  for (MachineBasicBlock *MBB : post_order(MF)) {
+    LiveIns[MBB].init(*TRI);
+    LiveOuts[MBB].init(*TRI);
+    WorkListSet.insert(MBB);
+    WorkList.push(MBB);
+  }
+  while (!WorkList.empty()) {
+    MachineBasicBlock *MBB = WorkList.front();
+    WorkList.pop();
+    WorkListSet.erase(MBB);
+    LivePhysRegs CurrentLive;
+    CurrentLive.init(*TRI);
+
+    // Calculates LiveOuts by iterating over each successor of the current MBB
+    // and adding each successor's LiveIns to the current MBB's LiveOuts.
+    // This ensures LiveOuts is the union of LiveIns of all successor MBBs.
+    for (MachineBasicBlock *Succ : MBB->successors()) {
+      for (MCPhysReg Reg : LiveIns[Succ]) {
+        LiveOuts[MBB].addReg(Reg);
```

Comment:
Nit: We could compute the LiveOuts in a final step, and keep that loop only about liveins computation.